pkg/reconciler/cluster/controller.go

```diff
@@ -36,63 +41,115 @@ const (
 	SyncerModeNone
 )
 
+const GVRForLocationInLogicalClusterIndexName = "GVRForLocationInLogicalCluster"
+
+func GetGVRForLocationInLogicalClusterIndexKey(location, clusterName string, gvr metav1.GroupVersionResource) string {
+	return location + "$$" + apiresource.GetClusterNameAndGVRIndexKey(clusterName, gvr)
+}
+
+const LocationInLogicalClusterIndexName = "LocationInLogicalCluster"
+
+func GetLocationInLogicalClusterIndexKey(location, clusterName string) string {
+	return location + "/" + clusterName
+}
+
 // NewController returns a new Controller which reconciles Cluster resources in the API
 // server it reaches using the REST client.
 //
 // When new Clusters are found, the syncer will be run there using the given image.
 func NewController(cfg *rest.Config, syncerImage string, kubeconfig clientcmdapi.Config, resourcesToSync []string, syncerMode SyncerMode) *Controller {
-	client := clusterv1alpha1.NewForConfigOrDie(cfg)
 	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
 	stopCh := make(chan struct{}) // TODO: hook this up to SIGTERM/SIGINT
 
 	crdClient := apiextensionsv1client.NewForConfigOrDie(cfg)
 
 	c := &Controller{
-		queue:           queue,
-		client:          client,
-		crdClient:       crdClient,
-		syncerImage:     syncerImage,
-		kubeconfig:      kubeconfig,
-		stopCh:          stopCh,
-		resourcesToSync: resourcesToSync,
-		syncerMode:      syncerMode,
-		syncers:         map[string]*Syncer{},
-	}
-
-	sif := externalversions.NewSharedInformerFactoryWithOptions(clusterclient.NewForConfigOrDie(cfg), resyncPeriod)
+		queue:             queue,
+		clusterClient:     typedcluster.NewForConfigOrDie(cfg),
+		apiResourceClient: typedapiresource.NewForConfigOrDie(cfg),
+		crdClient:         crdClient,
+		syncerImage:       syncerImage,
+		kubeconfig:        kubeconfig,
+		stopCh:            stopCh,
+		resourcesToSync:   resourcesToSync,
+		syncerMode:        syncerMode,
+		syncers:           map[string]*syncer.Syncer{},
+		apiImporters:      map[string]*APIImporter{},
+	}
+
+	sif := externalversions.NewSharedInformerFactoryWithOptions(versionedclient.NewForConfigOrDie(cfg), resyncPeriod)
 	sif.Cluster().V1alpha1().Clusters().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
 		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
 		DeleteFunc: func(obj interface{}) { c.deletedCluster(obj) },
 	})
-	c.indexer = sif.Cluster().V1alpha1().Clusters().Informer().GetIndexer()
+	c.clusterIndexer = sif.Cluster().V1alpha1().Clusters().Informer().GetIndexer()
+
+	enqueueRelatedCluster := func(obj interface{}) {
```

Comment:
This feels like it should be a method on the Controller type. TODO it at least and we can clean it up later.

Reply:
Fixed in commit 6a4dcd07a4a268ef7dd0158950e2a1c8452b9eb3